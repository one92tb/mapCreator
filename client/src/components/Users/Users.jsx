import React from "react";
import { connect } from "react-redux";
import { fetchUsers } from "../../actions/user/fetchUsers";
import { changePermissions } from "../../actions/user/changePermissions";
import { deleteAccount } from "../../actions/user/deleteAccount";
import PropTypes from "prop-types";
import {
  Wrapper,
  Select,
  Input,
  Form,
  TableContainer,
  Table,
  Thead,
  Tbody,
  Tr,
  Th,
  Td,
  RemoveBtn,
  Option
} from "./style";

class Users extends React.Component {
  componentDidMount() {
    const { fetchUsers } = this.props;
    fetchUsers();
  }

  handleChangeUser = () => {
    
  }

  handleChangeStatus = (e, user) => {
    const { changePermissions } = this.props;
    const status = e.target.value === "Admin" ? true : false;
    changePermissions(status, user.id);
  };

  removeUser = id => {
    const { deleteAccount } = this.props;
    deleteAccount(id);
  };

  render() {
    const { users } = this.props;
    return (
      <Wrapper>
        <Form>
          <Input
            onChange={this.handleChange}
            type="text"
            name="city"
            placeholder="search user"
          />
        </Form>
        <TableContainer>
          <Table>
            <Thead>
              <tr>
                <Th>id</Th>
                <Th>name</Th>
                <Th>change status</Th>
                <Th>remove</Th>
              </tr>
            </Thead>
            <Tbody>
              {users.map((user, id) => {
                return (
                  <Tr key={user.id}>
                    <Td>{id + 1}</Td>
                    <Td>{user.login}</Td>
                    <Td>
                      {user.id === 1 ? (
                        "Admin"
                      ) : (
                        <Select onChange={e => this.handleChangeStatus(e, user)}>
                          <Option key={id}>
                            {user.isAdmin ? "Admin" : "User"}
                          </Option>
                          <Option key={id + 1}>
                            {user.isAdmin ? "User" : "Admin"}
                          </Option>
                        </Select>
                      )}
                    </Td>
                    <Td>
                      {user.id !== 1 && (
                        <RemoveBtn
                          onClick={id => this.removeUser(user.id)}
                          src={"delete.png"}
                        />
                      )}
                    </Td>
                  </Tr>
                );
              })}
            </Tbody>
          </Table>
        </TableContainer>
      </Wrapper>
    );
  }
}

const mapDispatchToProps = {
  fetchUsers,
  changePermissions,
  deleteAccount
};

const mapStateToProps = state => ({
  users: state.user.users
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Users);

Users.propTypes = {
  users: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      login: PropTypes.string.isRequired,
      password: PropTypes.string.isRequired,
      isAdmin: PropTypes.bool.isRequired
    })
  ),
  fetchUsers: PropTypes.func.isRequired,
  changePermissions: PropTypes.func.isRequired,
  deleteAccount: PropTypes.func.isRequired
};
