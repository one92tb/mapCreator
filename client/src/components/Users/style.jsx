import styled from "styled-components";
import { css } from "styled-components";

const Wrapper = styled.div`
  background: #f2f2f2;
  height: 100%;
  padding: 40px 20px;
`;

const Select = styled.select`
  width: 150px;
  height: 30px;
  color: #000;
  display: block;
  text-align: center;
  border-radius: 5px;
  background: #fff;
  text-align: center;
  text-align-last: center;
  margin-left: auto;
  margin-right: auto;
`;

const Input = styled.input`
  height: 40px;
  width: 200px;
  text-align: center;
  border-radius: 5px;
  border: 1px solid transparent;
  margin-right: 50px;
`;

const Form = styled.form`
  height: calc(10% - 20px);
  margin-bottom: 20px;
  display: flex;
  justify-content: flex-end;
  background: #00b8e6;
  align-items: center;
  border-radius: 3px;
`;

const TableContainer = styled.div`
  height: 90%;
`;

const Table = styled.table`
  width: 100%;
  display: block;
  height: 100%;
`;

const Thead = styled.thead`
  display: table;
  width: 100%;
`;

const Tbody = styled.tbody`
  overflow: auto;
  overflow-x: hidden;
  display: block;
  width: 100%;
  height: 100%;
  height: calc(100% - 48px);
  text-align: center;
`;

const Tr = styled.tr`
  display: table;
  width: 100%;
  table-layout: fixed;
  margin: 5px 0;
`;

const Cell = css`
  padding: 0.75rem;
  text-align: center;
  width: 10%;

  &:not(:first-child) {
    width: 15%;
  }
`;

const Th = styled.th`
  ${Cell};
  color: #fff;
  background: #00b8e6;
`;

const Td = styled.td`
  ${Cell};
  background: #4ddbff;
`;

const RemoveBtn = styled.img`
  &:hover{
    cursor: pointer;
  }
`

export {
  Wrapper,
  Select,
  Input,
  Form,
  TableContainer,
  Table,
  Thead,
  Tbody,
  Tr,
  Th,
  Td,
  RemoveBtn
};
