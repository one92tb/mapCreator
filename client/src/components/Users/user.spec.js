import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { Users } from "./Users";

describe("user component", () => {
  let wrapper;

  it("passess all props to users component", () => {
    const fetchUsers = jest.fn();
    const changePermissions = jest.fn();
    const deleteAccount = jest.fn();
    const users = [
      {
        id: 1,
        isAdmin: true,
        login: "user1",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      },
      {
        id: 2,
        isAdmin: false,
        login: "user2",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      }
    ];

    wrapper = shallow(
      <Users
        fetchUsers={fetchUsers}
        changePermissions={changePermissions}
        deleteAccount={deleteAccount}
        users={users}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("componentDidMount", () => {
    const fetchUsers = jest.fn();
    const changePermissions = jest.fn();
    const deleteAccount = jest.fn();
    const users = [
      {
        id: 1,
        isAdmin: true,
        login: "user1",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      },
      {
        id: 2,
        isAdmin: false,
        login: "user2",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      }
    ];

    wrapper = shallow(
      <Users
        fetchUsers={fetchUsers}
        changePermissions={changePermissions}
        deleteAccount={deleteAccount}
        users={users}
      />
    );

    expect(fetchUsers).toHaveBeenCalledTimes(1);
  });

  it("handle change status - from user to admin", () => {
    const fetchUsers = jest.fn();
    const changePermissions = jest.fn();
    const deleteAccount = jest.fn();
    const users = [
      {
        id: 1,
        isAdmin: true,
        login: "user1",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      },
      {
        id: 2,
        isAdmin: false,
        login: "user2",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      }
    ];

    wrapper = shallow(
      <Users
        fetchUsers={fetchUsers}
        changePermissions={changePermissions}
        deleteAccount={deleteAccount}
        users={users}
      />
    );

    const select = wrapper.find("select");

    select.simulate("change", {
      target: {
        name: "user",
        value: "Admin"
      }
    });

    expect(changePermissions).toHaveBeenCalledTimes(1);
    expect(changePermissions).toHaveBeenCalledWith(true, 2);
  });

  it("handle change status - from admin to user", () => {
    const fetchUsers = jest.fn();
    const changePermissions = jest.fn();
    const deleteAccount = jest.fn();
    const users = [
      {
        id: 1,
        isAdmin: true,
        login: "user1",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      },
      {
        id: 2,
        isAdmin: true,
        login: "user2",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      }
    ];

    wrapper = shallow(
      <Users
        fetchUsers={fetchUsers}
        changePermissions={changePermissions}
        deleteAccount={deleteAccount}
        users={users}
      />
    );

    const select = wrapper.find("select");

    select.simulate("change", {
      target: {
        name: "user",
        value: "User"
      }
    });

    expect(changePermissions).toHaveBeenCalledTimes(1);
    expect(changePermissions).toHaveBeenCalledWith(false, 2);
  });

  it("handle change user - change state", () => {
    const fetchUsers = jest.fn();
    const changePermissions = jest.fn();
    const deleteAccount = jest.fn();
    const users = [
      {
        id: 1,
        isAdmin: true,
        login: "user1",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      },
      {
        id: 2,
        isAdmin: true,
        login: "user2",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      }
    ];

    wrapper = shallow(
      <Users
        fetchUsers={fetchUsers}
        changePermissions={changePermissions}
        deleteAccount={deleteAccount}
        users={users}
      />
    );

    const input = wrapper.find("input");

    input.simulate("change", {
      target: {
        name: "user",
        value: "user2"
      }
    });

    expect(wrapper.state().login).toBe("user2");
  });

  it("handle change user - change state", () => {
    const fetchUsers = jest.fn();
    const changePermissions = jest.fn();
    const deleteAccount = jest.fn();
    const users = [
      {
        id: 1,
        isAdmin: true,
        login: "user1",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      },
      {
        id: 2,
        isAdmin: true,
        login: "user2",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      }
    ];

    wrapper = shallow(
      <Users
        fetchUsers={fetchUsers}
        changePermissions={changePermissions}
        deleteAccount={deleteAccount}
        users={users}
      />
    );

    const removeBtn = wrapper.find("img");
    removeBtn.simulate("click", 2);
    expect(deleteAccount).toHaveBeenCalledTimes(1);
    expect(deleteAccount).toHaveBeenCalledWith(2);
  });

  it("map and filter users", () => {
    const fetchUsers = jest.fn();
    const changePermissions = jest.fn();
    const deleteAccount = jest.fn();
    const users = [
      {
        id: 1,
        isAdmin: true,
        login: "user1",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      },
      {
        id: 2,
        isAdmin: true,
        login: "user2",
        password: "$2b$10$4YBAEGFh9uAbQBBS1INaSuY3vS9yMuaxQaQ2oBaDu1IpWmNrMkYYK"
      }
    ];

    wrapper = shallow(
      <Users
        fetchUsers={fetchUsers}
        changePermissions={changePermissions}
        deleteAccount={deleteAccount}
        users={users}
      />
    );

    const tbody = wrapper.find("tbody");
    console.log(tbody.html());

    wrapper.setState({
      login: "user1"
    });

    const filteredTr = tbody.children().filterWhere(tr => {
      return (
        tr
          .childAt(1)
          .children()
          .text() === wrapper.instance().state.login
      );
    });

    expect(filteredTr.length).toBe(1);
  });
});
