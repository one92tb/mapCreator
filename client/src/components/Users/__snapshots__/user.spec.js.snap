// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user component passess all props to users component 1`] = `
<div>
  <form>
    <input
      name="user"
      onChange={[Function]}
      placeholder="search user"
      type="text"
    />
  </form>
  <div>
    <table>
      <thead>
        <tr>
          <th>
            id
          </th>
          <th>
            name
          </th>
          <th>
            change status
          </th>
          <th>
            remove
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="1"
        >
          <td>
            1
          </td>
          <td>
            user1
          </td>
          <td>
            Admin
          </td>
          <td />
        </tr>
        <tr
          key="2"
        >
          <td>
            2
          </td>
          <td>
            user2
          </td>
          <td>
            <select
              name="user"
              onChange={[Function]}
            >
              <option
                key="1"
              >
                User
              </option>
              <option
                key="2"
              >
                Admin
              </option>
            </select>
          </td>
          <td>
            <img
              onClick={[Function]}
              src="delete.png"
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
